numba_dpex/core/runtime/_dpexrt_python.c

```diff
@@ -1269,6 +1273,97 @@ static PyObject *DPEXRT_sycl_queue_to_python(queuestruct_t *queuestruct)
+ * eventstruct_t instance.
+ *
+ * @param    obj            A dpctl.SyclEvent Python object
+ * @param    event_struct   An instance of the struct numba-dpex uses to
+ *                          represent a dpctl.SyclEvent inside Numba.
+ * @return   {return}       Return code indicating success (0) or failure (-1).
+ */
+static int DPEXRT_sycl_event_from_python(PyObject *obj,
+                                         eventstruct_t *event_struct)
+{
+
+    struct PySyclEventObject *event_obj = NULL;
+    DPCTLSyclEventRef event_ref = NULL;
+    PyGILState_STATE gstate;
+
+    // We are unconditionally casting obj to a struct PySyclEventObject*. If
+    // the obj is not a struct PySyclEventObject* then the SyclEvent_GetEventRef
+    // will error out.
+    event_obj = (struct PySyclEventObject *)obj;
+
+    DPEXRT_DEBUG(
+        drt_debug_print("DPEXRT-DEBUG: In DPEXRT_sycl_event_from_python.\n"););
+
+    if (!(event_ref = SyclEvent_GetEventRef(event_obj))) {
+        DPEXRT_DEBUG(drt_debug_print(
+            "DPEXRT-ERROR: SyclEvent_GetEventRef returned NULL at "
+            "%s, line %d.\n",
+            __FILE__, __LINE__));
+        goto error;
+    }
+
+    event_struct->parent = obj;
+    event_struct->event_ref = event_ref;
+
+    return 0;
+
+error:
+    // If the check failed then decrement the refcount and return an error
+    // code of -1.
+    // Decref the Pyobject of the array
+    // ensure the GIL
+    DPEXRT_DEBUG(drt_debug_print(
+        "DPEXRT-ERROR: Failed to unbox dpctl SyclEvent into a Numba "
+        "eventstruct at %s, line %d\n",
+        __FILE__, __LINE__));
+
+    return -1;
+}
+
+/*!
+ * @brief A helper function that boxes a Numba-dpex eventstruct_t object into a
+ * dctl.SyclEvent PyObject using the eventstruct_t's parent attribute.
+ *
+ * If there is no parent pointer stored in the eventstruct, then an error will
+ * be raised.
+ *
+ * @param    eventstruct    A Numba-dpex eventstruct object.
+ * @return   {return}       A PyObject created from the eventstruct->parent, if
+ *                          the PyObject could not be created return NULL.
+ */
+static PyObject *DPEXRT_sycl_event_to_python(eventstruct_t *eventstruct)
+{
+    PyObject *orig_event = NULL;
+    PyGILState_STATE gstate;
+
+    orig_event = eventstruct->parent;
+    // FIXME: Better error checking is needed to enforce the boxing of the event
+    // object. For now, only the minimal is done as the returning of SyclEvent
+    // from a dpjit function should not be a used often and the dpctl C API for
+    // type checking etc. is not ready.
+    if (orig_event == NULL) {
+        PyErr_Format(PyExc_ValueError,
+                     "In 'box_from_eventstruct_parent', "
+                     "failed to create a new dpctl.SyclEvent object.");
+        return NULL;
+    }
+
+    DPEXRT_DEBUG(
+        drt_debug_print("DPEXRT-DEBUG: In DPEXRT_sycl_event_to_python.\n"););
+
+    Py_INCREF(orig_event);
```

Comment:
Can you add a comment for the `Py_INCREF` call to explain why it is called and issues if not done. Will save problems in future.